lib/internal/crypto/mac.js

```diff
@@ -0,0 +1,209 @@
+  kSignJobModeVerify,
+} = internalBinding('crypto');
+
+const {
+  getHashLength,
+  hasAnyNotIn,
+  jobPromise,
+  lazyDOMException,
+  normalizeHashName,
+  validateBitLength,
+  validateKeyOps,
+  kHandle,
+  kKeyObject,
+} = require('internal/crypto/util');
+
+const {
+  codes: {
+    ERR_MISSING_OPTION,
+    ERR_INVALID_ARG_TYPE,
+  }
+} = require('internal/errors');
+
+const {
+  generateKey,
+} = require('internal/crypto/keygen');
+
+const {
+  CryptoKey,
+  SecretKeyObject,
+  createSecretKey,
+  isKeyObject,
+} = require('internal/crypto/keys');
+
+async function hmacGenerateKey(algorithm, extractable, keyUsages) {
+  const { hash, name } = algorithm;
+  let { length } = algorithm;
+  if (hash === undefined)
+    throw new ERR_MISSING_OPTION('algorithm.hash');
+
+  if (length === undefined)
+    length = getHashLength(hash.name);
+
+  validateBitLength(length, 'algorithm.length', true);
+
+  const usageSet = new SafeSet(keyUsages);
+  if (hasAnyNotIn(usageSet, 'sign', 'verify')) {
+    throw lazyDOMException(
+      'Unsupported key usage for an HMAC key',
+      'SyntaxError');
+  }
+  return new Promise((resolve, reject) => {
+    generateKey('hmac', { length }, (err, key) => {
+      if (err) {
+        return reject(lazyDOMException(
+          'The operation failed for an operation-specific reason',
+          'OperationError'));
+      }
+
+      resolve(new CryptoKey(
+        key,
+        { name, length, hash: { name: hash.name } },
+        ArrayFrom(usageSet),
+        extractable));
+    });
+  });
+}
+
+async function hmacImportKey(
+  format,
+  keyData,
+  algorithm,
+  extractable,
+  keyUsages) {
+  const { hash } = algorithm;
+  if (hash === undefined)
+    throw new ERR_MISSING_OPTION('algorithm.hash');
+
+  const usagesSet = new SafeSet(keyUsages);
+  if (hasAnyNotIn(usagesSet, 'sign', 'verify')) {
+    throw lazyDOMException(
+      'Unsupported key usage for an HMAC key',
+      'SyntaxError');
+  }
+  let keyObject;
+  switch (format) {
+    case 'node.keyObject': {
+      if (!isKeyObject(keyData))
+        throw new ERR_INVALID_ARG_TYPE('keyData', 'KeyObject', keyData);
+
+      if (keyData.type !== 'secret') {
+        throw lazyDOMException(
+          `Unable to import HMAC key with format ${format}`);
+      }
+
+      keyObject = keyData;
+      break;
+    }
+    case 'raw': {
+      const len = keyData.byteLength * 8;  // Number of bits
+
+      let { length: checkLength } = algorithm;
+      if (checkLength === undefined)
+        checkLength = getHashLength(algorithm.hash.name);
+
+      if (len !== checkLength) {
+        // The Web Crypto spec actually states:
+        //   If the algorithm.length is greater than length, DataError
+        //   Or, If the algorithm.length <= length - 8, DataError
+        // Here we enforce a stricter check since we require
+        // algorithm.length to be a multiple of 8.
+        throw lazyDOMException('Key length mismatch', 'DataError');
+      }
```

Comment:
> // The Web Crypto spec actually states:
> //   If the algorithm.length is greater than length, DataError
> //   Or, If the algorithm.length <= length - 8, DataError
> // Here we enforce a stricter check since we require
> // algorithm.length to be a multiple of 8.

Does it really? Maybe just for generation?

This check (even a laxer one) is definitely not in place in any of the major browsers - i've tested both chrome and firefox the snippet below works on all of them.

```
await crypto.subtle.importKey('raw', new Uint8Array(1), { hash: 'SHA-256', name: 'HMAC' }, true, ['sign'])
```

HMAC keys are the only ones where implementers can get any binary data represented as a CryptoKey so that it can be wrapped using wrapKey ops.

Reply:
@jasnell actually, the problem here was that the length was enforced even when the "length" optional algorithm property was not provided.

I believe this should be in place and checked - IF - `{ hash: 'SHA-256', name: 'HMAC', **length** }` length is present.

Reply:
Ah right. Digging back into the spec I noticed that I missed the zero length check also. Added!